src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -327,40 +327,88 @@ bool fx_muxer_t::resolve_hostpolicy_dir(host_mode_t mode,
     return false;
 }
 
-void fx_muxer_t::resolve_roll_forward(pal::string_t& fx_dir, const pal::string_t& fx_ver, const fx_ver_t& specified)
+void fx_muxer_t::resolve_roll_forward(pal::string_t& fx_dir,
+    const pal::string_t& fx_ver,
+    const fx_ver_t& specified,
+    const bool& patch_roll_fwd,
+    const bool& no_candidate_fx_roll_fwd)
 {
     trace::verbose(_X("Attempting FX roll forward starting from [%s]"), fx_ver.c_str());
 
     std::vector<pal::string_t> list;
     pal::readdir(fx_dir, &list);
     fx_ver_t most_compatible = specified;
-    for (const auto& version : list)
+    if (!specified.is_prerelease())
     {
-        trace::verbose(_X("Inspecting version... [%s]"), version.c_str());
-        fx_ver_t ver(-1, -1, -1);
-        if (!specified.is_prerelease() && fx_ver_t::parse(version, &ver, true) && // true -- only prod. prevents roll forward to prerelease.
-            ver.get_major() == specified.get_major() &&
-            ver.get_minor() == specified.get_minor())
+        if (no_candidate_fx_roll_fwd)
         {
-            // Pick the greatest production that differs only in patch.
-            most_compatible = std::max(ver, most_compatible);
+            trace::verbose(_X("'No candidate fx roll forward' enabled. Looking for the least production greater than or equal to [%s]"), fx_ver.c_str());
+            fx_ver_t next_lowest(-1, -1, -1);
+            for (const auto& version : list)
+            {
+                fx_ver_t ver(-1, -1, -1);
+                if (fx_ver_t::parse(version, &ver, true) &&
+                    (ver > specified || ver == specified))
+                {
+                    next_lowest = (next_lowest == fx_ver_t(-1, -1, -1)) ? ver : std::min(next_lowest, ver);
+                }
+            }
+            if (next_lowest == fx_ver_t(-1, -1, -1))
+            {
+                trace::verbose(_X("No production greater than or equal to [%s] found."), fx_ver.c_str());
+            }
+            else
+            {
+                trace::verbose(_X("Found version [%s]"), next_lowest.as_str().c_str());
+                most_compatible = next_lowest;
```

Comment:
We should update _most_compatible_ only if the requested version is missing. Take the example of 2.0.0 and 2.0.1 present on the machine, with NoCandidateRollFX enabled but PatchRollForward disabled and requested version as 2.0.0.

With your change above, _most_compatible_ becomes 2.0.1 and app executes against 2.0.1 even though patch roll forward was disabled. We should be running against 2.0.0.